File: sdks/sandbox/python/src/opensandbox/adapters/converter/exception_converter.py
#
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""
Exception converter utilities.

Provides conversion functions from API exceptions to domain exceptions,
similar to the Kotlin SDK ExceptionConverter pattern.

This module handles:
1. Converting openapi-python-client generated exceptions
2. Converting httpx HTTP errors
3. Converting network/IO errors
4. Parsing error response bodies to extract SandboxError information
"""

import json
import logging
from datetime import timedelta
from http import HTTPStatus
from typing import Any

from httpx import (
    ConnectError,
    ConnectTimeout,
    HTTPStatusError,
    NetworkError,
    PoolTimeout,
    ReadTimeout,
    TimeoutException,
    TransportError,
    WriteTimeout,
)

from opensandbox.api.execd.errors import UnexpectedStatus as ExecdUnexpectedStatus
from opensandbox.api.lifecycle.errors import (
    UnexpectedStatus as LifecycleUnexpectedStatus,
)
from opensandbox.exceptions import (
    InvalidArgumentException,
    SandboxApiException,
    SandboxConnectionException,
    SandboxError,
    SandboxException,
    SandboxInternalException,
    SandboxRateLimitException,
    SandboxTimeoutException,
)
from opensandbox.transport._decision import parse_retry_after

logger = logging.getLogger(__name__)

UNEXPECTED_STATUS_TYPES = (LifecycleUnexpectedStatus, ExecdUnexpectedStatus)


class ExceptionConverter:
    """
    Exception converter utilities following Kotlin SDK patterns.

    Provides static methods to convert various exceptions to sandbox exceptions,
    including proper parsing of error response bodies.
    """

    @staticmethod
    def to_sandbox_exception(e: Exception) -> SandboxException:
        """
        Convert any exception to a SandboxException.

        Following Kotlin SDK pattern:
        - SandboxException -> return as-is
        - API client exceptions -> convert to SandboxApiException
        - IOError/network errors -> convert to SandboxInternalException with network message
        - IllegalArgumentError/ValueError -> convert to SandboxInternalException with usage message
        - Other exceptions -> convert to SandboxInternalException with unexpected error message

        Args:
            e: The original exception

        Returns:
            A SandboxException subclass
        """
        # If already a SandboxException, return as-is
        if isinstance(e, SandboxException):
            return e

        # Handle openapi-python-client UnexpectedStatus error
        if _is_unexpected_status_error(e):
            return _convert_unexpected_status_to_api_exception(e)

        # Handle httpx HTTPStatusError
        if _is_httpx_status_error(e):
            return _convert_httpx_error_to_api_exception(e)

        # httpx connection failures (DNS, TCP connect, TLS handshake,
        # connect timeout) — pre-send, semantically "cannot reach".
        # ConnectTimeout must be dispatched before generic TimeoutException
        # because it is a subclass of both TimeoutException and ConnectError.
        if isinstance(e, (ConnectTimeout, ConnectError, NetworkError)):
            return SandboxConnectionException(
                message=f"Network connectivity error: {e}",
                cause=e,
                is_retryable=True,
            )

        # httpx timeout family (read, write, pool, or the synthetic
        # ReadTimeout raised by the retry wrapper when overall_deadline
        # is exhausted) — the request was sent but did not finish in time.
        if isinstance(e, (ReadTimeout, WriteTimeout, PoolTimeout, TimeoutException)):
            return SandboxTimeoutException(
                message=f"Request timed out: {e}",
                cause=e,
            )

        # OS-level connectivity errors (DNS, socket errors, etc.) not
        # already routed through httpx's typed hierarchy.
        if isinstance(e, (ConnectionError,)):
            return SandboxConnectionException(
                message=f"Network connectivity error: {e}",
                cause=e,
            )
        if isinstance(e, (IOError, OSError)):
            return SandboxInternalException(
                message=f"Network connectivity error: {e}",
                cause=e,
            )

        # Any other httpx TransportError (opaque low-level failures).
        if isinstance(e, TransportError):
            return SandboxConnectionException(
                message=f"Network connectivity error: {e}",
                cause=e,
            )

        # Handle validation and argument errors (SDK usage errors)
        # - ValueError/TypeError are typically raised for invalid user inputs or model validation
        # - Pydantic ValidationError represents invalid input data for SDK models
        try:
            from pydantic import ValidationError  # type: ignore

            if isinstance(e, ValidationError):
                return InvalidArgumentException(message=str(e), cause=e)
        except Exception:
            # If pydantic isn't available for some reason, just ignore and continue
            pass

        if isinstance(e, (ValueError, TypeError)):
            return InvalidArgumentException(message=str(e), cause=e)

        # Handle unsupported operations
        if isinstance(e, NotImplementedError):
            return SandboxInternalException(
                message=f"Operation not supported: {e}",
                cause=e,
            )

        # Default to unexpected error
        return SandboxInternalException(
            message=f"Unexpected SDK error occurred: {e}",
            cause=e,
        )


def _is_unexpected_status_error(e: Exception) -> bool:
    """Check if exception is an openapi-python-client UnexpectedStatus error."""
    return isinstance(e, UNEXPECTED_STATUS_TYPES)


def _is_httpx_status_error(e: Exception) -> bool:
    """Check if exception is an httpx HTTPStatusError."""
    return isinstance(e, HTTPStatusError)


def _retry_after_from_headers(headers: Any) -> timedelta | None:
    if not headers:
        return None
    try:
        raw = headers.get("Retry-After") or headers.get("retry-after")
    except Exception:
        return None
    return parse_retry_after(raw if isinstance(raw, str) else None)


# Status codes the SDK treats as transient (the default idempotent
# retryable set from RetryPolicy). Surfaced via ``is_retryable`` so
# callers get a machine-checkable retry decision even after the SDK has
# exhausted its own retries. Budget/deadline/cancellation are the only
# things that would force this to False; those paths raise a timeout
# rather than an API exception.
_RETRYABLE_STATUS_CODES = frozenset(
    {
        HTTPStatus.TOO_MANY_REQUESTS,   # 429
        HTTPStatus.BAD_GATEWAY,         # 502
        HTTPStatus.SERVICE_UNAVAILABLE, # 503
    }
)


def _build_api_exception(
    *,
    status_code: int,
    content: bytes | None,
    cause: Exception,
    request_id: str | None = None,
    retry_after: timedelta | None = None,
) -> SandboxApiException:
    """Build a Sandbox(ApiException|RateLimitException) from raw fields."""
    from opensandbox.adapters.converter.response_handler import (
        _raw_body_message_fragment,
    )

    sandbox_error = _parse_error_body(content) if content else None
    message = f"API error: HTTP {status_code}"
    if sandbox_error is not None and sandbox_error.code != SandboxError.UNEXPECTED_RESPONSE:
        if sandbox_error.message:
            message = f"{message}: {sandbox_error.message}"
    else:
        # Unstructured body: splice the raw response body (truncated) into the
        # message so logs carry the server's own explanation instead of only
        # "API error: HTTP 400". The full body stays on ``response_body``.
        # Structured codes with an empty message are preserved on the error.
        raw_fragment = _raw_body_message_fragment(content)
        if raw_fragment:
            message = f"{message}: {raw_fragment}"
            if sandbox_error is None or sandbox_error.code == SandboxError.UNEXPECTED_RESPONSE:
                sandbox_error = SandboxError(SandboxError.UNEXPECTED_RESPONSE, raw_fragment)
    is_retryable = status_code in _RETRYABLE_STATUS_CODES
    if status_code == HTTPStatus.TOO_MANY_REQUESTS:
        return SandboxRateLimitException(
            message=message,
            status_code=status_code,
            cause=cause,
            error=sandbox_error,
            request_id=request_id,
            retry_after=retry_after,
            response_body=content if isinstance(content, bytes) else None,
            is_retryable=is_retryable,
        )
    return SandboxApiException(
        message=message,
        status_code=status_code,
        cause=cause,
        error=sandbox_error,
        request_id=request_id,
        response_body=content if isinstance(content, bytes) else None,
        is_retryable=is_retryable,
    )


def _convert_unexpected_status_to_api_exception(e: Exception) -> SandboxApiException:
    """Convert openapi-python-client UnexpectedStatus to SandboxApiException."""
    status_code = getattr(e, "status_code", 0)
    content = getattr(e, "content", b"")
    # openapi-python-client's UnexpectedStatus does not carry headers,
    # so Retry-After / request_id are not recoverable here.
    return _build_api_exception(
        status_code=int(status_code) if status_code else 0,
        content=content if isinstance(content, bytes) else None,
        cause=e,
    )


def _convert_httpx_error_to_api_exception(e: Exception) -> SandboxApiException:
    """Convert httpx HTTPStatusError to SandboxApiException."""
    response = getattr(e, "response", None)
    status_code = response.status_code if response else 0
    content = response.content if response else b""
    request_id = None
    retry_after = None
    if response is not None:
        from opensandbox.adapters.converter.response_handler import extract_request_id

        request_id = extract_request_id(response.headers)
        retry_after = _retry_after_from_headers(response.headers)

    return _build_api_exception(
        status_code=int(status_code) if status_code else 0,
        content=content if isinstance(content, bytes) else None,
        cause=e,
        request_id=request_id,
        retry_after=retry_after,
    )


def _parse_error_body(body: Any) -> SandboxError | None:
    """
    Parse error body to extract SandboxError information.

    Similar to Kotlin SDK's parseSandboxError function.

    Args:
        body: The error response body (bytes, str, or dict)

    Returns:
        SandboxError if parsing succeeds, None otherwise
    """
    if body is None:
        return None

    try:
        # Convert bytes to string
        if isinstance(body, bytes):
            if not body:
                return None
            body = body.decode("utf-8", errors="replace")

        if isinstance(body, str) and not body:
            return None

        # Parse JSON string
        if isinstance(body, str):
            try:
                body = json.loads(body)
            except json.JSONDecodeError:
                # If not JSON, return error with the raw string as message
                return SandboxError(
                    code=SandboxError.UNEXPECTED_RESPONSE,
                    message=body,
                )

        # FastAPI HTTPException bodies are commonly wrapped as {"detail": {"code": ..., "message": ...}}.
        if isinstance(body, dict):
            if isinstance(body.get("detail"), dict):
                body = body["detail"]

            code: str | None = body.get("code")
            message: str | None = body.get("message")

            if code:
                return SandboxError(code=code, message=message or "")

        return None

    except Exception as ex:
        logger.debug(f"Failed to parse error body: {ex}")
        return None


def parse_sandbox_error(body: Any) -> SandboxError | None:
    """
    Public function to parse error body to SandboxError.

    Exposed for use by other modules that need to parse error bodies.
    """
    return _parse_error_body(body)

File: sdks/sandbox/kotlin/sandbox/src/main/kotlin/com/alibaba/opensandbox/sandbox/infrastructure/adapters/converter/ExceptionConverter.kt
/*
 * Copyright 2025 Alibaba Group Holding Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.opensandbox.sandbox.infrastructure.adapters.converter

import com.alibaba.opensandbox.sandbox.api.infrastructure.ClientError
import com.alibaba.opensandbox.sandbox.api.infrastructure.ClientException
import com.alibaba.opensandbox.sandbox.api.infrastructure.ServerError
import com.alibaba.opensandbox.sandbox.api.infrastructure.ServerException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxApiException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxConnectionException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxError
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxError.Companion.UNEXPECTED_RESPONSE
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxInternalException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxRateLimitException
import com.alibaba.opensandbox.sandbox.domain.exceptions.SandboxTimeoutException
import com.alibaba.opensandbox.sandbox.transport.RetryDeadlineExceededException
import com.alibaba.opensandbox.sandbox.transport.RetryPolicy
import com.alibaba.opensandbox.sandbox.transport.parseRetryAfter
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import okhttp3.Response
import java.io.IOException
import java.io.InterruptedIOException
import java.net.ConnectException
import java.net.NoRouteToHostException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.time.Duration
import javax.net.ssl.SSLException
import com.alibaba.opensandbox.sandbox.api.diagnostic.infrastructure.ClientError as DiagnosticClientError
import com.alibaba.opensandbox.sandbox.api.diagnostic.infrastructure.ClientException as DiagnosticClientException
import com.alibaba.opensandbox.sandbox.api.diagnostic.infrastructure.ServerError as DiagnosticServerError
import com.alibaba.opensandbox.sandbox.api.diagnostic.infrastructure.ServerException as DiagnosticServerException
import com.alibaba.opensandbox.sandbox.api.execd.infrastructure.ApiResponse as ExecdApiResponse
import com.alibaba.opensandbox.sandbox.api.execd.infrastructure.ClientError as ExecdClientError
import com.alibaba.opensandbox.sandbox.api.execd.infrastructure.ClientException as ExecdClientException
import com.alibaba.opensandbox.sandbox.api.execd.infrastructure.ServerError as ExecdServerError
import com.alibaba.opensandbox.sandbox.api.execd.infrastructure.ServerException as ExecdServerException

/**
 * Returns `true` when this throwable represents an expected "file or directory does not exist"
 * outcome rather than a genuine failure.
 *
 * Detection is intentionally restricted to the explicit [SandboxError.FILE_NOT_FOUND] server
 * error code rather than a bare HTTP 404. A 404 whose body cannot be parsed is mapped to
 * [SandboxError.UNEXPECTED_RESPONSE] and may indicate a real endpoint/routing/configuration
 * regression, which must stay loud (ERROR) instead of being silently downgraded.
 *
 * Callers (and the adapters themselves) use this to avoid treating a missing file as an error,
 * e.g. logging it at ERROR level with a full stack trace, which is just noise for a perfectly
 * normal control-flow case such as polling for a not-yet-created file.
 */
fun Throwable.isFileNotFound(): Boolean = this is SandboxApiException && error.code == SandboxError.FILE_NOT_FOUND

private fun buildSandboxApiException(
    message: String?,
    statusCode: Int,
    cause: Throwable? = null,
    errorBody: Any? = null,
    requestId: String? = null,
    retryAfter: Duration? = null,
): SandboxApiException {
    val sandboxError =
        parseSandboxError(errorBody) ?: when {
            statusCode == 429 -> SandboxError(SandboxError.RATE_LIMIT, message)
            errorBody is String -> SandboxError(UNEXPECTED_RESPONSE, errorBody)
            else -> SandboxError(UNEXPECTED_RESPONSE)
        }
    val responseBody = errorBody as? String
    val isRetryable = statusCode in RetryPolicy.DEFAULT_IDEMPOTENT_STATUS

    if (statusCode == 429) {
        return SandboxRateLimitException(
            message = message,
            statusCode = statusCode,
            cause = cause,
            error = sandboxError,
            requestId = requestId,
            retryAfter = retryAfter,
            responseBody = responseBody,
            isRetryable = isRetryable,
        )
    }

    return SandboxApiException(
        message = message,
        statusCode = statusCode,
        cause = cause,
        error = sandboxError,
        requestId = requestId,
        responseBody = responseBody,
        isRetryable = isRetryable,
    )
}

/**
 * Build the public API exception surface for handwritten OkHttp adapter paths.
 *
 * The generated clients flow through [toSandboxException]; direct OkHttp calls
 * must expose the same raw body, request metadata, rate-limit subtype, and
 * retryability signal.
 */
fun Response.toSandboxApiException(
    responseBody: String? = body?.string(),
    message: (statusCode: Int, responseBody: String?) -> String,
): SandboxApiException =
    buildSandboxApiException(
        message = message(code, responseBody),
        statusCode = code,
        errorBody = responseBody,
        requestId = header("X-Request-ID"),
        retryAfter = parseRetryAfter(header("Retry-After")),
    )

/** Build the same error surface from an execd generated-client HTTP response. */
internal fun ExecdApiResponse<*>.toSandboxApiException(message: (statusCode: Int, responseBody: String?) -> String): SandboxApiException {
    val errorBody: Any? =
        when (this) {
            is ExecdClientError<*> -> body
            is ExecdServerError<*> -> body
            else -> null
        }
    val responseBody = errorBody as? String
    return buildSandboxApiException(
        message = message(statusCode, responseBody),
        statusCode = statusCode,
        errorBody = errorBody,
        requestId = headers.extractRequestId(),
        retryAfter = headers.extractRetryAfter(),
    )
}

fun Exception.toSandboxException(): SandboxException {
    return when (this) {
        is SandboxException -> this
        is ClientException, is ServerException,
        is ExecdClientException, is ExecdServerException,
        is DiagnosticClientException, is DiagnosticServerException,
        -> this.toApiException()
        // The retry interceptor exhausted the overall deadline before the next
        // attempt: surface as a timeout, never retryable.
        is RetryDeadlineExceededException ->
            SandboxTimeoutException(
                message = "Request timed out: ${this.message}",
                cause = this,
                isRetryable = false,
            )
        // Pre-send connectivity failures: DNS, TCP connect, TLS handshake.
        is ConnectException, is UnknownHostException, is NoRouteToHostException, is SSLException ->
            SandboxConnectionException(
                message = "Network connectivity error: ${this.message}",
                cause = this,
                isRetryable = true,
            )

        // Read timeouts: the request was sent but did not finish in time.
        is SocketTimeoutException ->
            if (this.message?.lowercase()?.contains("connect") == true) {
                SandboxConnectionException(
                    message = "Network connectivity error: ${this.message}",
                    cause = this,
                    isRetryable = true,
                )
            } else {
                SandboxTimeoutException(
                    message = "Request timed out: ${this.message}",
                    cause = this,
                    isRetryable = false,
                )
            }
        is InterruptedIOException ->
            SandboxTimeoutException(
                message = "Request timed out: ${this.message}",
                cause = this,
                isRetryable = false,
            )
        is IOException ->
            SandboxInternalException(
                message = "Network connectivity error: ${this.message}",
                cause = this,
            )
        is IllegalStateException, is IllegalArgumentException ->
            SandboxInternalException(
                message = "SDK internal usage error: ${this.message}",
                cause = this,
            )
        is UnsupportedOperationException ->
            SandboxInternalException(
                message = "Operation not supported: ${this.message}",
                cause = this,
            )
        else ->
            SandboxInternalException(
                message = "Unexpected SDK error occurred: ${this.message}",
                cause = this,
            )
    }
}

private fun Exception.toApiException(): SandboxApiException {
    val (statusCode, rawResponse) =
        when (this) {
            is ClientException -> this.statusCode to this.response
            is ServerException -> this.statusCode to this.response
            is ExecdClientException -> this.statusCode to this.response
            is ExecdServerException -> this.statusCode to this.response
            is DiagnosticClientException -> this.statusCode to this.response
            is DiagnosticServerException -> this.statusCode to this.response
            else -> 0 to null
        }

    val headers: Map<String, List<String>>? =
        when (rawResponse) {
            is ClientError<*> -> rawResponse.headers
            is ServerError<*> -> rawResponse.headers
            is ExecdClientError<*> -> rawResponse.headers
            is ExecdServerError<*> -> rawResponse.headers
            is DiagnosticClientError<*> -> rawResponse.headers
            is DiagnosticServerError<*> -> rawResponse.headers
            else -> null
        }

    val requestId = headers?.extractRequestId()

    val errorBody =
        when (rawResponse) {
            is ClientError<*> -> rawResponse.body
            is ExecdServerError<*> -> rawResponse.body
            is ServerError<*> -> rawResponse.body
            is ExecdClientError<*> -> rawResponse.body
            is DiagnosticClientError<*> -> rawResponse.body
            is DiagnosticServerError<*> -> rawResponse.body
            else -> null
        }

    return buildSandboxApiException(
        message = this.message,
        statusCode = statusCode,
        cause = this,
        errorBody = errorBody,
        requestId = requestId,
        retryAfter = headers?.extractRetryAfter(),
    )
}

private fun Map<String, List<String>>.extractRequestId(): String? {
    return entries.firstOrNull { (key, _) ->
        key.equals("X-Request-ID", ignoreCase = true)
    }?.value?.firstOrNull()?.takeIf { it.isNotBlank() }
}

private fun Map<String, List<String>>.extractRetryAfter(): java.time.Duration? {
    val raw =
        entries.firstOrNull { (key, _) ->
            key.equals("Retry-After", ignoreCase = true)
        }?.value?.firstOrNull()
    return parseRetryAfter(raw)
}

fun parseSandboxError(body: Any?): SandboxError? {
    if (body == null) return null

    return runCatching {
        val jsonElement: JsonElement =
            when (body) {
                is String -> jsonParser.parseToJsonElement(body)
                else -> jsonParser.encodeToJsonElement(body)
            }

        val generic = jsonParser.decodeFromJsonElement<GenericErrorBody>(jsonElement)

        if (!generic.code.isNullOrBlank()) {
            SandboxError(code = generic.code, message = generic.message)
        } else {
            null
        }
    }.getOrNull()
}

@Serializable
private data class GenericErrorBody(
    val code: String? = null,
    val message: String? = null,
)
